File: LeetCode75_LevelTwo/Task_Scheduler/Task_Scheduler.cc
// 621. Task Scheduler C++
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        if (n == 0) return tasks.size();
        int freq[26] = {0};
        for (char c: tasks) freq[c-'A']++;
        sort(freq, freq+26);
        int max_val = freq[25]-1;
        int idle_slots = max_val * n;
        for (int i = 24; i >= 0; i--) {
            idle_slots -= min(freq[i], max_val);
        }
        return (idle_slots > 0) ? tasks.size() + idle_slots: tasks.size();
    }
};

File: LeetCode75_LevelTwo/Task_Scheduler/Task_Scheduler.py
# 621. Task Scheduler Python

class Solution(object):
    def leastInterval(self, tasks, n):
        """
        :type tasks: List[str]
        :type n: int
        :rtype: int
        """
        freq = [0] * 26
        for t in tasks:
            freq[ord(t) - ord('A')] += 1
        freq.sort()

        f_max = freq.pop()
        idle_time = (f_max - 1) * n

        while freq and idle_time > 0:
            idle_time -= min(f_max - 1, freq.pop())
        idle_time = max(0, idle_time)

        return idle_time + len(tasks)
